### ISSUE TITLE
[ignore-http-errors] Add option to ignore/skip download errors

### ISSUE THREAD

**Issue by deepakjois** [coursera-dl/coursera-dl#453]
I am running this on Ubuntu 14.04, with the default version of Python 3 inside a virtualenv

```
coursera-dl -n  -- understanding-financial-markets
```

It runs but gets stuck trying to download a PDF file from citeseerx.

```
Downloading: understanding-financial-markets/03_other-financial-markets/02_alternative-assets/08_alternative-versus-traditional-investments_download?doi=10.1.1.579.7843&rep=rep1&type=pdf
Downloading http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.579.7843&rep=rep1&type=pdf -> understanding-financial-markets/03_other-financial-markets/02_alternative-assets/08_alternative-versus-traditional-investments_download?doi=10.1.1.579.7843&rep=rep1&type=pdf
Starting new HTTP connection (1): citeseerx.ist.psu.edu
```

The problem is with citeseerx, but there appears to be no way of instructing coursera-dl to skip and move on. Are there any workarounds or solutions to the problem?

**Comment by deepakjois** [comment]
Looks like this is related to #443. I was able to apply the workaround `--formats "srt mp4 pdf"` to get around this problem.

However, it would be nice to have a timeout or some option to skip the resources so that it does not get stuck at a specific URL.

**Comment by balta2ar** [comment]
> it would be nice to have a timeout or some option to skip the resources so that it does not get stuck at a specific URL.

Agree.

As an alternative solution, try using `--aria`. This downloader is configurable and more robust to errors.

**Comment by deepakjois** [comment]
I did try using the `--aria` option. In this particular case, it gaves the same result. Is it possible to configure aria on the coursera-dl commandline, or do you do it some other way?

**Comment by balta2ar** [comment]
Hm. coursera-dl appears to have no options (I may be wrong) to control aria. However, can always tune `~/.aria2/aria2.conf` instead.

**Comment by deepakjois** [comment]
Ok, I will try that. Feel free to close this issue, if you think it’s no longer an issue.

**Comment by balta2ar** [comment]
I've renamed the issue. I've always wanted that feature and it's missing indeed (but it's present in edx-dl, for example).

**Comment by deepakjois** [comment]
I just tried with the following settings in `aria.conf`:
```
console-log-level=debug
timeout=2
```

It still does not work as expected (full log below except a few bits hidden for privacy):

```

03/14 16:00:19 [INFO] <<--- --- --- ---

03/14 16:00:19 [INFO]   --- --- --- ---

03/14 16:00:19 [INFO]   --- --- --- --->>

03/14 16:00:19 [INFO] aria2 1.18.1 x86_64-unknown-linux-gnu

03/14 16:00:19 [INFO] Logging started.

03/14 16:00:19 [INFO] Checking configured addresses

03/14 16:00:19 [INFO] Not considered: 127.0.0.1

03/14 16:00:19 [INFO] Found configured address: xxx

03/14 16:00:19 [INFO] Not considered: ::1

03/14 16:00:19 [INFO] Not considered: xxx

03/14 16:00:19 [INFO] IPv4 configured=1, IPv6 configured=0
03/14 16:00:19 [INFO] 173 certificate(s) were imported.

03/14 16:00:19 [DEBUG] 1 RequestGroup(s) added.

03/14 16:00:19 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0

03/14 16:00:19 [DEBUG] Selected from normCands

03/14 16:00:19 [DEBUG] FeedbackURISelector selected http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.579.7843&rep=rep1&type=pdf

03/14 16:00:19 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0

03/14 16:00:19 [INFO] CUID#6 - Resolving hostname citeseerx.ist.psu.edu

03/14 16:00:19 [DEBUG] Failed to delete socket event:Bad file descriptor

03/14 16:00:19 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0

03/14 16:00:19 [INFO] CUID#6 - Name resolution complete: citeseerx.ist.psu.edu -> 130.203.136.95

03/14 16:00:19 [INFO] CUID#6 - Connecting to 130.203.136.95:80
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:20 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:22 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:23 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:24 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:25 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:26 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:27 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:28 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:29 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:30 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
[#429470 0B/0B CN:1 DL:0B]
[#429470 0B/0B CN:1 DL:0B]
03/14 16:00:21 [DEBUG] CUID#6 - socket: read:0, write:0, hup:0, err:0
…
```

And this goes on.

**Comment by deepakjois** [comment]
I have been downloading a bunch of courses, and this set of options in my `~/.aria2/aria.conf` have been helpful:

```
connect-timeout=2
timeout=2
bt-stop-timeout=1
```

Just putting it here for the record.

**Comment by imaia** [comment]
Just got myself a "no route to host" here from a broken link to pdf file. A feature to ignore download errors would be indeed nice.

**Comment by GlowingCrystallineEntity** [comment]
Would like to see this too...

Just got a "resources" URL that just goes to a book-for-sale page, so it's not really anything downloadable, but the URL ends with "detail.aspx?ID=..." ... and coursera-dl is trying to create a filename with the question mark (?), which fails (on Windows).

**Comment by GlowingCrystallineEntity** [comment]
... and as another use-case, I just got another course that references an entire project on (of all things) GitHub, which fails since coursera-dl doesn't have my login info (for starters).  And there's little reason to set that up, since I can d/l project zips from github.

**Comment by rbrito** [comment]
Yes, I believe that we need to, somehow, prune the download from sites that are not coursera's (yes, I know that there are very valuable resources that instructors put on their own sites or that they put for papers on other sites).

We just have to decide what to prune and what not. One initial heuristic would be to ignore anything that is the root of an entire website.

**Comment by GlowingCrystallineEntity** [comment]
Things that resolve to HTML (aspx, etc.) files probable aren't that useful to download, but what would be nice is if the program kept a list of all failed/skipped downloads and printed that before it exited (maybe optionally to a file), so that users could easily look through that list and handle the links themselves, with confidence that they aren't missing anything.

**Comment by rbrito** [comment]
Just got another nasty link pointing to a site that is paywalled:

http://ieeexplore.ieee.org/xpl/login.jsp?tp=&arnumber=5128907&url=http%3A%2F%2Fieeexplore.ieee.org%2Fiel5%2F69%2F5173046%2F05128907.pdf%3Farnumber%3D5128907

We should inspect our URLs to see if their base domain is ieeexplore... I don't like the idea of maintaining a blacklist, but this one is perhaps safe to use, unless the person happens to be using coursera-dl under a verified IP address (say, a university).

OK, back to work now that I have to submit work and time is running out. I just decided to write, since I just stumbled upon it.
